File: bak-emacs/cscope/cscope-indexer.py
#Created on 2009-4-4

#@author: tiger
#'''
import getopt
import sys
import os
import re

def usage( ):
    print( "cscope-indexer [ -v ] [-f database_file ] [-i list_file ] [ -l ] [ -r ]" )
    pass
    

def listRecurse( rootDir, reg, fileName ):
    fp = open( fileName, 'w' )
    for root, dirs, files in os.walk( rootDir ):
        for file in files :
            if reg.search(file) != None :
                fp.write( os.path.join( root, file ) )
    fp.close()
    pass

def lisrNoRecurse( rootDir, reg, fileName ):
    fp = open( fileName, 'w' )
    for file in os.listdir( rootDir ) :
        if reg.search(file) != None :
            fp.write( os.path.join( rootDir, file ) )
    fp.close()
    pass


if __name__ == '__main__':
    databaseFile = "cscope.out"
    listFile = "cscope.files"
    listOnly = False
    recurse = False
    verbose = False
    dir = "."
    try:
        opts,args=getopt.getopt(sys.argv[1:],'vf:i:lr')
        for flag,data in opts :
            if flag == "-v":
                verbose = True
            elif flag == "-f":
                if data == "":
                    usage()
                    os.system.exit()
                else:
                    databaseFile = data
            elif flag == "-i":
                if data == "":
                    usage()
                    os.system.exit()
                else:
                    listFile = data
            elif flag == "-l":
                listOnly = True
            elif flag == "-r":
                recurse = True
            else:
                usage()
                os.system.exit()
        if len(args) != 0:
            dir = args[0]
    except getopt.GetoptError:
        usage()
        os.system.exit()
    if verbose == True:
        print( "Creating list of files to index ..." ) 
    os.chdir( dir )
    reg = re.compile('\.([chly](xx|pp)*|cc|hh)$')
    if recurse == False :
        lisrNoRecurse( dir, reg, listFile )
    else:
        listRecurse( dir, reg, listFile )
    if verbose == True:
        print( "Creating list of files to index ... done" )
    if listOnly == True :
        os.system.exit()
    if verbose == True :
        print( "Indexing files ..." )
    os.system( "cscope -b -i "+listFile+" -f "+databaseFile )
    if verbose == True :
        print( "Indexing files ... done" )
    
    pass

File: bak-emacs/cscope/cscope-indexer.pl
#!/usr/bin/perl -w
# cscope-indexer.pl --- 
# Author: Ye Wenbin @gmail.com>
# Created: 10 Sep 2008
# Version: 0.01
use warnings;
use strict;
use Getopt::Long qw/:config pass_through/;
use File::Find;
# May have to edit this:
if ( $^O =~ /win/ ) {
   $ENV{PATH} = "d:\\Programs\\cygwin\\usr\\local\\bin;$ENV{PATH}";
}
else {
   $ENV{PATH} = "/usr/local/bin:/sbin:/usr/sbin:/bin:/usr/bin:$ENV{PATH}";
}
my $LIST_ONLY;
my $DIR='.';
my $LIST_FILE='cscope.files';
my $DATABASE_FILE='cscope.out';
my $RECURSE;
my $VERBOSE;
GetOptions(
   "file=s" => \$DATABASE_FILE,
   "index=s" => \$LIST_FILE,
   "list" => \$LIST_ONLY,
   "recurse" => \$RECURSE,
   "verbose" => \$VERBOSE,
);
$DIR = shift if @ARGV;
message("Creating list of files to index ...");
my @files;
if ( $RECURSE ) {
   find( sub {
             if ( -f $_ || -l $_ ) {
                 push @files, $File::Find::name;
             }
         }, $DIR);
} else {
   @files = ;
}
@files = sort map {s{^\./}{}; $_} grep { /\.([chly](xx|pp)?|cc|hh)$/
                            && $_ !~ m{/CVS/}
                                && $_ !~ m{/RCS/}
                            } @files;
open(my $fh, ">", $LIST_FILE) or die "Can't create file $LIST_FILE: $!";
print {$fh} join("\n", @files), "\n";
close $fh;
message("Creating list of files to index ... done");
if ( $LIST_ONLY ) {
   exit;
}
message("Indexing files ...");
system("cscope", "-b", "-i", $LIST_FILE, "-f", $DATABASE_FILE) == 0 or die "$!";
message("Indexing files ... done");
sub message {
   print @_, "\n" if $VERBOSE;
}